Enhanced MQTT Data for Serial Battery Configuration

swingstate opened the pull request:
I'm excited to share my first pull request, which focuses on enriching the MQTT data for batteries in a serial configuration. This enhancement includes not only additional data fields but also improves the decimal accuracy for both live view displayed data and MQTT transmissions.

### Changes Made:

**New Data Fields for HomeAssistant MQTT Auto Discovery:**

1. Midpoint Voltage:
- Added midpoint voltage information to provide a more comprehensive view of the battery's state in a serial configuration.

2. Midpoint Deviation:
- Introduced the midpoint deviation metric, offering insights into the balance and stability of the battery pack.

3. Additional Metrics:
- Consumed Ampere Hours: Tracking the total ampere hours consumed over time.
- Instantaneous Power: Providing real-time power consumption data for more immediate insights.
- Charged Energy: Monitoring the total energy charged into the battery.
- Discharged Energy: Keeping tabs on the energy discharged from the battery.

**Decimal Accuracy Improvement:**
To enhance precision, the decimal accuracy for both live view and MQTT-transmitted Smartshunt data has been increased.

schlimmchen commented on src/BatteryStats.cpp:
```diff
@@ -41,7 +41,7 @@ void BatteryStats::getLiveViewData(JsonVariant& root) const
     root[F("data_age")] = getAgeSeconds();
 
-    addLiveViewValue(root, "SoC", _SoC, "%", 0);
+    addLiveViewValue(root, "SoC", _SoC, "%", 1);
```
This is too generic. The JK BMS (and maybe the Pylontech as well) do not report tens of percents. I suggest you introduce a protected "SoC resolution" variable in the base class BatteryStats, initialize it to zero, and overwrite it to be 1 in VictronSmartShuntStats.

swingstate replied:
maybe later, as this goes beyond my skills. I'll revert this one first.

schlimmchen commented on src/BatteryStats.cpp:
```diff
@@ -378,4 +382,9 @@ void VictronSmartShuntStats::mqttPublish() const {
+    MqttSettings.publish(F("battery/instantaneousPower"), String(_instantaneousPower));
+    MqttSettings.publish(F("battery/consumpedAmpHours"), String(_consumedAmpHours));
 }
+
```
Please avoid adding empty lines like this one.

schlimmchen commented on src/MqttHandlePylontechHass.cpp:
```diff
@@ -62,6 +62,12 @@ void MqttHandlePylontechHassClass::publishConfig()
     publishSensor("Charge voltage (BMS)", NULL, "settings/chargeVoltage", "voltage", "measurement", "V");
     publishSensor("Charge current limit", NULL, "settings/chargeCurrentLimitation", "current", "measurement", "A");
     publishSensor("Discharge current limit", NULL, "settings/dischargeCurrentLimitation", "current", "measurement", "A");
+    publishSensor("SmartShunt battery midpoint", NULL, "midpointVoltage", "voltage", "measurement", "V");
+    publishSensor("SmartShunt battery midpoint Deviation", NULL, "midpointDeviation", "battery", "measurement", "%");
+    publishSensor("SmartShunt Instantaneous Power", NULL, "instantaneousPower", "power", "measurement", "W");
+    publishSensor("SmartShunt Consumed AmpHours", NULL, "consumpedAmpHours", "energy", "measurement", "Ah");
+    publishSensor("SmartShunt Total Charged Energy", NULL, "chargedEnergy", "energy", "total_increasing", "kWh");
+    publishSensor("SmartShunt Total DisCharged Energy", NULL, "dischargedEnergy", "energy", "total_increasing", "kWh");
```
We may leave it like that for now, however, this is not the way forward. You are now publishing "SmartShunt" sensors to HomeAssistent, even if the actual battery is a Pylontech or a JK BMS. This is not yet fixed in general. The whole file needs to be refactored such that only the appropriate sensors are published. Your change makes it worse, though...

swingstate replied:
I am aware of this flaw and piggy bagging on the Pylontech, since I haven't found any other way to get the sensor published.

swingstate replied:
Though I do not agree that this is making things worse as this is just using existing code strategies / structures and extending on it.

schlimmchen reviewed (changes requested):
Thanks for contributing and congrats to your first pull request :rocket: Please have a look at my review comments.

swingstate commented on lib/VeDirectFrameHandler/VeDirectShuntController.h:
```diff
@@ -12,30 +12,32 @@ class VeDirectShuntController : public VeDirectFrameHandler {
-        int32_t H1;                     // Depth of the deepest discharge
-        int32_t H2;                     // Depth of the last discharge
-        int32_t H3;                     // Depth of the average discharge
+        float_t H1;                     // Depth of the deepest discharge
+        float_t H2;                     // Depth of the last discharge
+        float_t H3;                     // Depth of the average discharge
         int32_t H4;                     // Number of charge cycles
         int32_t H5;                     // Number of full discharges
         int32_t H6;                     // Cumulative Amp Hours drawn
-        int32_t H7;                     // Minimum main (battery) voltage
-        int32_t H8;                     // Maximum main (battery) voltage
+        float_t H7;                     // Minimum main (battery) voltage
+        float_t H8;                     // Maximum main (battery) voltage
         int32_t H9;                     // Number of seconds since last full charge
         int32_t H10;                    // Number of automatic synchronizations
         int32_t H11;                    // Number of low main voltage alarms
         int32_t H12;                    // Number of high main voltage alarms
         int32_t H13;                    // Number of low auxiliary voltage alarms
         int32_t H14;                    // Number of high auxiliary voltage alarms
-        int32_t H15;                    // Minimum auxiliary (battery) voltage
-        int32_t H16;                    // Maximum auxiliary (battery) voltage
-        int32_t H17;                    // Amount of discharged energy
-        int32_t H18;                    // Amount of charged energy
+        float_t H15;                    // Minimum auxiliary (battery) voltage
+        float_t H16;                    // Maximum auxiliary (battery) voltage
+        float_t H17;                    // Amount of discharged energy
+        float_t H18;                    // Amount of charged energy
+        float_t VM;                     // Mid-point voltage of the battery bank
+        float_t DM;                     // Mid-point deviation of the battery bank
```
Okay, understood. The latter is what I am after, respectively I want to use float since they (from my understanding) provide higher accuracy.

swingstate replied:
I will revert the changes here and look into increasing accuracy / BatteryStats separately, once this PR is done.

swingstate commented:
@schlimmchen I made couple of adjustments based on your feedback and deleted the .DS files (at least I think I deleted them). Can you please take a look. 

I merged the last updated from the main repo into my dev and now have 2 onBattery PCBs running with the latest (main + my changes) updates. So far things look good.

schlimmchen commented:
@swingstate This is much better, thanks!

This changeset now removes three of those (other) `.DS_Store` files that were previously there. The goal is to have a changeset that only does what is needed to get the feature going.

Regarding the issue that you publish your values alongside the Pylontech sensors you wrote:

> Though I do not agree that this is making things worse as this is just using existing code strategies / structures and extending on it.

I disagree. People with Pylonthech batteries would see sensors that don't even show values for their Battery, as they are specific to the SmartShunt.

Since you prepared the PR, at least one more thing changed: The use of the "F" macro. That would have to be removed. And you would have to rebase your work onto the current state of the development branch.

I finished the Home Assistent integration for the JK BMS in #640. Your work and mine now conflict. However, the work I did will solve the issue with the missing distinction of the battery provider.

If you are okay with it, I can offer to help this change along and prepare a respective new PR that is based on #640. I would use the work you did to put it in the right new places and prepare clean commits. I could try to preserve your authorship, however I cannot promise that it will work as I intend.

swingstate commented:
I will not capacity to finish and test this in the near future. Maybe you @schlimmchen or @LukasK13 can look into this and/or merge this with the effort Lukas is looking into for the MQTT Powerlimiter settings.

ottelo9 commented:
If you need someone to test it (if ready), I could do it. I have a Pylontech battery.

schlimmchen commented:
@swingstate @fellpower Could you please test the firmware built as the result of [this PR's build run](https://github.com/helgeerbe/OpenDTU-OnBattery/actions/runs/9316611082?pr=573) against a SmartShunt? We are interested to see:
1. Does the live view show midpoint voltage and deviation?
2. Is the midpoint voltage and deviation published to MQTT?
3. In Home Assistent, are the midpoint voltage and deviation auto-discovered?
4. Are the values the same in every place and are they plausible?

fellpower commented:
I can confirm - yet - that it seems to be working

![image](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/73786796/303520df-1535-40d0-b33f-06c6fca3ef24)

I have no deviation, so i have to make one and test it - but the voltage is correct ^^

schlimmchen commented:
Thanks, @fellpower, very nice. However, there is an issue: The midpoint value should not read 13.00 V if it is 13.3x V. I think I know where I messed up. I will provide an updated firmware later today.

fellpower commented:
No, we have to thank YOU ;)

Ill test, if u are ready. i tested today the deviation precentage, its working as it should. i can see clearly the deviatition, if u have one.

schlimmchen commented:
@fellpower The new build run is ready with [new firmware](https://github.com/helgeerbe/OpenDTU-OnBattery/actions/runs/9322634481?pr=573). This version should display the midpoint voltage with two decimal places.

Looking forward to see the screenshot confirming success :wink:

fellpower commented:
Seems okay for me ;)

![image](https://github.com/helgeerbe/OpenDTU-OnBattery/assets/73786796/83409c2b-7357-40f8-bdcc-03b89dac6288)

schlimmchen commented:
@fellpower Nice, thanks for quickly helping out testing this feature!
